pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/performance/xml/AvoidNonRestrictiveQueries.xml

```diff
@@ -0,0 +1,110 @@
+        <description>Query with where and limit statement</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Something {
+    public static void main( String[] as ) {
+        Account[] accs = [ select id from account where id = 1 limit 1];
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Subquery without limit</description>
+        <expected-problems>1</expected-problems>
+        <expected-messages>
+            <message>Avoid SOQL sub-queries without a where or limit statement</message>
+        </expected-messages>
+        <code><![CDATA[
+public class Something {
+    public static void main( String[] as ) {
+        Account[] accs = [ select id, (SELECT Id FROM Contact) from account where id = 1];  //Bad
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Multiple non restrictive queries</description>
+        <expected-problems>3</expected-problems>
+        <expected-linenumbers>3,4,5</expected-linenumbers>
+        <expected-messages>
+            <message>Avoid SOQL queries without a where or limit statement</message>
+            <message>Avoid SOQL queries without a where or limit statement</message>
+            <message>Avoid SOQL queries without a where or limit statement</message>
+        </expected-messages>
+        <code><![CDATA[
+public class Something {
+    public static Account[] main( String[] as ) {
+        Account[] accs = [ select id from account ];
+        Account[] accs2 = [ select id from account ];
+        return [SELECT Id FROM Contact];
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Restrictive queries</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Something {
+    public static Account[] main( String[] as ) {
+        Account[] accs = [ select id from account where name='foo' and id < 100 LIMIT 5 ];
+        Account[] accs2 = [ select id from account where id > 1 ];
+        return [SELECT Id FROM Contact limit 1];
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>subquery with limit, main query without</description>
+        <expected-problems>1</expected-problems>
+        <expected-messages>
+            <message>Avoid SOQL queries without a where or limit statement</message>
+        </expected-messages>
+        <code><![CDATA[
+class Query {
+    void method() {
+        Account[] accounts = [ select id, (SELECT Id FROM Contact where id = 1 limit 1) from account ];
+    }
+}
+]]></code>
+    </test-code>
+
+    <test-code disabled="true">
+        <description>main query with limit, subquery without is OK</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+class Query {
+    void method() {
+        Account[] accounts = [ select id, (SELECT Id FROM Contact) from account limit 1 ];
+    }
+}
```

Comment:
@dschach
I've updated the original rule from #663 to see what's possible. I now have some contradictions, maybe you can help?
Should this query be flagged by the rule, because it will exceed governor limits?
Or maybe should we flag this, because the subquery SELECT potentially selects all rows and is therefore a performance issue? (the rule is currently in category performance, so this could make sense...).

This test case basically contradicts the test case in lines 42-45...

Reply:
@adangel If I were to put the rule in plain English:
1. In the SOQL query, throw away any subqueries - we don't care about those.
2. On the main query, require either the word WHERE or LIMIT.
Therefore, the query here is perfectly fine.

Reply:
Thanks for the clarification - that makes it easier. The rule should only check the main query.